### PR TITLE
CDD-1919 and CDD-1957: GET Unpulled and pulled notifications list

### PR DESCRIPTION
**Pull request by rashadmughal**

### CODE REVIEW

File: app/uk/gov/hmrc/apinotificationpull/controllers/EnhancedNotificationsController.scala [hunk 1]
```diff
@@ -47,8 +47,22 @@ class EnhancedNotificationsController @Inject()(enhancedApiNotificationQueueServ
...
+  private def getList(notificationStatus: NotificationStatus.Value): Action[AnyContent] =
+  (headerValidator.validateAcceptHeader andThen headerValidator.validateXClientIdHeader).async { implicit request =>
+
+    implicit val hc: HeaderCarrier = buildHeaderCarrier(request)
+
+    enhancedApiNotificationQueueService.getAllNotificationsBy(notificationStatus).map { notifications =>
+        Ok(enhancedXmlBuilder.toXml(notifications, notificationStatus)).as(XML)
+    } recover recovery
+  }
+
```
Comment (googley42) [inline app/uk/gov/hmrc/apinotificationpull/controllers/EnhancedNotificationsController.scala]: what we are doing with 
```scala
          case Right(n) => Result(
            header = ResponseHeader(OK),
            body = HttpEntity.Strict(ByteString(n.payload), n.headers.get(CONTENT_TYPE)))
            .withHeaders(n.headers.toSeq: _*)
          case Left(nfe: NotFoundException) =>
            logger.info(s"Notification not found for id: $notificationId", nfe)
            ErrorNotFound.XmlResult
          case Left(bre: BadRequestException) =>
            logger.info(s"$badRequestText for id: $notificationId", bre)
            errorBadRequest(badRequestText).XmlResult
          case Left(e) =>
            logger.error(s"Internal server error for notification id: $notificationId", e)
            ErrorInternalServerError.XmlResult
        }.recover(recovery)
```
is packing an exception as a value/model in the connector, then decoding/unpacking it here as an error result.
We might as well return an error type of `ErrorResponse` from our connector, and also do the error logging there. Then in the controller all we need to do when we get a `Left` is `errorResult.Xml` to get the error response

File: app/uk/gov/hmrc/apinotificationpull/connectors/EnhancedApiNotificationQueueConnector.scala [hunk 2]
```diff
@@ -31,6 +30,10 @@ class EnhancedApiNotificationQueueConnector @Inject()(config: ServiceConfigurati
 
   private lazy val serviceBaseUrl: String = config.baseUrl("api-notification-queue")
 
+  def getAllNotificationsBy(notificationStatus: NotificationStatus.Value)(implicit hc: HeaderCarrier): Future[Notifications] = {
+    http.GET[Notifications](s"$serviceBaseUrl/notifications/${notificationStatus.toString}")
+  }
+
```
Comment (googley42) [inline app/uk/gov/hmrc/apinotificationpull/connectors/EnhancedApiNotificationQueueConnector.scala]: I think we should return a type of `Future[Either[ErrorResponse, Notification]]`

File: test/unit/connectors/EnhancedApiNotificationQueueConnectorSpec.scala [hunk 3]
```diff
@@ -113,5 +134,25 @@ class EnhancedApiNotificationQueueConnectorSpec extends UnitSpec with MockitoSug
...
+    "return an empty list of unpulled notifications when the downstream returns an empty list" in new Setup {
+
+      when(mockHttpClient.GET[Notifications](meq(s"http://api-notification-queue.url/notifications/unpulled"))
+        (any[HttpReads[Notifications]](), any[HeaderCarrier](), any[ExecutionContext])).thenReturn(Future.successful(Notifications(List())))
+
+      val result = await(enhancedApiNotificationQueueConnector.getAllNotificationsBy(Unpulled))
+
+      result shouldBe Notifications(List())
+    }
+
```
Comment (googley42) [inline test/unit/connectors/EnhancedApiNotificationQueueConnectorSpec.scala]: I think this should be `return an empty list of previously pulled notifications when the downstream returns an empty list`. Also other tests names that mention `pulled` should follow suit.

Comment (rashadmughal) [reply test/unit/connectors/EnhancedApiNotificationQueueConnectorSpec.scala]: :+1:

**Pull request merged**